universal-application-tool-0.0.1/app/repository/ApplicationRepository.java

```
@@ -5,62 +5,169 @@
+              .findList();
+      for (Application application : oldApplications) {
+        // Delete any in-progress drafts, and mark obsolete any old applications.
+        if (application.getLifecycleStage().equals(LifecycleStage.DRAFT)) {
+          application.setLifecycleStage(LifecycleStage.DELETED);
+        } else {
+          application.setLifecycleStage(LifecycleStage.OBSOLETE);
+        }
+        application.save();
+      }
+      Application application = new Application(applicant, program, LifecycleStage.ACTIVE);
+      application.save();
+      ebeanServer.commitTransaction();
+      return application;
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
+  private CompletionStage<Optional<Application>> perform(
+      long applicantId, long programId, Function<ApplicationArguments, Application> fn) {
     CompletionStage<Optional<Applicant>> applicantDb =
         applicantRepository.lookupApplicant(applicantId);
     CompletionStage<Optional<Program>> programDb = programRepository.lookupProgram(programId);
-    return applicantDb.thenCombineAsync(
-        programDb,
-        (applicantMaybe, programMaybe) -> {
-          if (applicantMaybe.isEmpty() || programMaybe.isEmpty()) {
-            return Optional.empty();
-          }
-          Applicant applicant = applicantMaybe.get();
-          Program program = programMaybe.get();
-          Application application = new Application(applicant, program, clock.instant());
-          application.save();
-          return Optional.of(application);
-        });
+    return applicantDb
+        .thenCombineAsync(
+            programDb,
+            (applicantMaybe, programMaybe) -> {
+              if (applicantMaybe.isEmpty()) {
+                throw new RuntimeException(new ApplicantNotFoundException(applicantId));
+              }
+              if (programMaybe.isEmpty()) {
+                throw new RuntimeException(new ProgramNotFoundException(programId));
+              }
```

Thought(non-blocking): this makes me feel that we should raise these exceptions in the corresponding repositories instead of here.